### PR TITLE
If Notifications service isn't registered, check if it could be activated

### PR DESCRIPTION
**Pull request by kbroulik**
DBus supports launching services on demand the first time a call is made to them.
In case the service isn't currently running, check whether it would be activated
when we try to call it.
Especially in case of session restoration the application might be launched before
the shell's notification service is fully up and running.

### CODE REVIEW

File: Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp [hunk 1]
```diff
@@ -316,13 +322,40 @@ const QDBusArgument &operator>>(const QDBusArgument &argument,
...
-		kObjectPath.utf16(),
-		kInterface.utf16()
-	).isValid();
+	enum Status {
+		Unknown = -1,
+		NotSupported,
+		Supported
+	};
+
+	static Status status = Unknown;
```
Comment (john-preston) [inline Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp]: Please look at how it is done here:

https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/platform/linux/specific_linux.cpp#L272-L294

Something like:
```
bool Supported() {
    static const auto Result = [] {
        if (blabla) {
            return true;
        }
        blabla;
        return false;
    }();
    return Result;
}
```

**Comment by john-preston** [comment]
@kbroulik And this really happens and the fix works? @ilya-fedin says he has such running-on-demand service and it works already, as if isValid() call activates it.

**Comment by ilya-fedin** [comment]
Yes, mate has a dbus-activable notification daemon and current code detects it.
Are there cases where this doesn't work? :thinking:

**Comment by kbroulik** [comment]
Odd. With my Telegram snap (1.9.9.4) it doesn't work, it just assumes no service was found and doesn't use native notifications in this case. With a test program with `QDBusInterface` (which one should never use) and a Telegram git build it launches the service as expected.

**Comment by ilya-fedin** [comment]
DE?

**Comment by kbroulik** [comment]
KDE Plasma. Oh, could maybe the snap sandbox be messing with dbus activation?

**Comment by ilya-fedin** [comment]
Maybe...

File: Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp [hunk 2]
```diff
@@ -316,13 +322,30 @@ const QDBusArgument &operator>>(const QDBusArgument &argument,
...
+			return true;
+		}
+
+		// Check if service would be activated when we call it
+		QDBusMessage message = QDBusMessage::createMethodCall(kDBusService.utf16(),
+			kDBusObjectPath.utf16(),
+			kDBusInterface.utf16(),
+			QStringLiteral("ListActivatableNames"));
+
+		QDBusMessage reply = QDBusConnection::sessionBus().call(message);
```
Comment (ilya-fedin) [inline Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp]: Maybe QDBusReply is better?
https://github.com/KDE/knotifications/blob/master/src/notifybypopup.cpp#L171

**Comment by kbroulik** [comment]
Ok, it appears to be a Snap sandboxing "issue".
The `Introspect` call `QDBusInterface` (which is why one should never use it) isn't allowed making the DBus-activation fail because effectively no call was ever made to the service.

**Pull request closed without merging**